File: pkg/data/render.go
package data

import (
	"fmt"
	"sort"
	"strings"
)

type DynamicFormRenderer interface {
	BeginForm()
	EndForm()
	BeginFieldset(label string)
	EndFieldset()
	Input(label, name, inputType, value string)
	Checkbox(label, name string, checked bool)
}

func Render(data any, r DynamicFormRenderer) error {
	r.BeginForm()
	defer r.EndForm()
	return renderValue(data, "", "", r)
}

func renderValue(val any, key, path string, r DynamicFormRenderer) error {
	switch v := val.(type) {
	case map[string]any:
		if key != "" {
			r.BeginFieldset(key)
		}

		keys := make([]string, 0, len(v))
		for k := range v {
			keys = append(keys, k)
		}
		sort.Strings(keys)

		for _, k := range keys {
			fullPath := joinPath(path, k)
			err := renderValue(v[k], k, fullPath, r)
			if err != nil {
				return err
			}
		}

		if key != "" {
			r.EndFieldset()
		}

	case string:
		r.Input(key, path, "text", v)
	case float64, float32, int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
		r.Input(key, path, "number", fmt.Sprintf("%v", v))
	case bool:
		r.Checkbox(key, path, v)
	default:
		return fmt.Errorf("at %v not handle value type %T", path, val)
	}
	return nil
}

func joinPath(parts ...string) string {
	nonEmpty := []string{}
	for _, part := range parts {
		if part != "" {
			nonEmpty = append(nonEmpty, part)
		}
	}
	return strings.Join(nonEmpty, ".")
}

File: front/src/pkg/data/render.ts
export interface DynamicFormRenderer {
  beginForm(): void;
  endForm(): void;
  beginObject(label: string): void;
  endObject(): void;
  input(label: string, name: string, inputType: string, value: string): void;
  checkbox(label: string, name: string, checked: boolean): void;
}

export function render(data: unknown, r: DynamicFormRenderer): Error | null {
  try {
    r.beginForm();
    renderValue(data, "", "", r);
    r.endForm();
    return null;
  } catch (e) {
    return e instanceof Error ? e : new Error(String(e));
  }
}

function renderValue(
  val: unknown,
  key: string,
  path: string,
  r: DynamicFormRenderer,
): void {
  // ------------------------
  // case: object / map[string]any
  // ------------------------
  if (isPlainObject(val)) {
    if (key !== "") {
      r.beginObject(key);
    }

    const entries = Object.entries(val);
    const sortedKeys = entries.map(([k]) => k).sort();

    for (const k of sortedKeys) {
      const nextVal = (val as Record<string, unknown>)[k];
      const fullPath = joinPath(path, k);
      renderValue(nextVal, k, fullPath, r);
    }

    if (key !== "") {
      r.endObject();
    }
    return;
  }

  // ------------------------
  // case: string
  // ------------------------
  if (typeof val === "string") {
    r.input(key, path, "text", val);
    return;
  }

  // ------------------------
  // case: numeric types (Go: float64, float32, int, etc.)
  // ------------------------
  if (typeof val === "number") {
    r.input(key, path, "number", String(val));
    return;
  }

  // ------------------------
  // case: boolean
  // ------------------------
  if (typeof val === "boolean") {
    r.checkbox(key, path, val);
    return;
  }

  // ------------------------
  // unsupported type
  // ------------------------
  throw new Error(`at ${path} not handle value type ${typeof val}`);
}

function joinPath(...parts: string[]): string {
  const nonEmpty = parts.filter((p) => p !== "");
  return nonEmpty.join(".");
}

function isPlainObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}
